File: Kakao/인턴_2020_수식 최대화.cpp
// 출제의도: "완전탐색(순열),문자열"

#include <string>
#include <vector>
#include <algorithm>
#include <iostream>

using namespace std;

vector<string>  parse(string expression) {
    vector<string>  parsed;
    int front=0;
    for(int i=0;i<expression.size();i++) {
        if(!isdigit(expression[i])) {
            parsed.push_back(expression.substr(front,i-front));
            parsed.push_back(expression.substr(i,1));
            front = i+1;
        }
    }
    parsed.push_back(expression.substr(front));
    return parsed;
}

void cal_infix(vector<string> &infix, string op) {
    for(int i=0;i<infix.size();i++) {
        long long result;
        if(infix[i]==op) {
            switch(op[0]) {
                case '+':
                    result = stoll(infix[i-1])+stoll(infix[i+1]);
                    break;
                case '-':
                    result = stoll(infix[i-1])-stoll(infix[i+1]);
                    break;
                case '*':
                    result = stoll(infix[i-1])*stoll(infix[i+1]);
                    break;
            }
            
            infix.erase(infix.begin()+i-1,infix.begin()+i+2);
            infix.insert(infix.begin()+i-1,to_string(result));
            i--;
        }
    }
}

long long solution(string expression) {
    long long answer = 0;
    vector<string>  parsed = parse(expression);
    vector<string> priority = {"*","+","-"};

    do {
        vector<string> infix = parsed;
        long long result=0;
        for(int i=0;i<priority.size();i++) {
            cal_infix(infix, priority[i]);
        }
        answer = max(answer, abs(stoll(infix[0])));
    } while(next_permutation(priority.begin(),priority.end()));
    return answer;
}

File: Kakao/인턴_2020_수식 최대화.swift
// "Divide and Conquer", Time Complexity: O(n), Space Complexity: O(n) (n: expression.size)

import Foundation

var precedence: [Character] = ["*", "+", "-"]

extension Array where Element == Character {
    mutating func next_permutation() -> Bool {
        var first: Index = 0, last: Index = self.count, i: Index = last-1
        
        while true {
            var i1: Index, i2: Index
            
            i1 = i
            i -= 1
            if self[i] < self[i1] {
                i2 = last
                repeat {
                    i2 -= 1
                } while !(self[i] < self[i2])
                
                self.swapAt(i, i2)
                
                let r = self[i1..<last].reversed()
                self.removeSubrange(i1..<last)
                self += r
                
                return true
            }
            
            guard i != first else { return false }
        }
    }
}

extension String {
    subscript(r: Range<Int>) -> String {
            let begin = self.index(self.startIndex, offsetBy: r.lowerBound)
            let end = self.index(self.startIndex, offsetBy: r.upperBound)
            return String(self[begin..<end])
    }
}

func calc(_ expression: String) -> Int64 {
    var result = Int64()
    var fnd = false

    for op in precedence.reversed() {
        guard !fnd else { break }

        for c in expression.enumerated().reversed() {
            guard !fnd && c.element == op else { continue }
            
            fnd.toggle()
            switch op {
                case "*":
                    result = calc(expression[0..<c.offset])*calc(expression[c.offset+1..<expression.count])
                case "+":
                    result = calc(expression[0..<c.offset])+calc(expression[c.offset+1..<expression.count])
                default:
                    result = calc(expression[0..<c.offset])-calc(expression[c.offset+1..<expression.count])
            }
        }
    }

    guard fnd else { return Int64(expression) ?? -1 }
    return result
}

func solution(_ expression:String) -> Int64 {
    var answer = Int64.min
    repeat {
        answer = max(answer, abs(calc(expression)))
    } while precedence.next_permutation()
    return answer
}
